### PR TITLE
Fix issue #5015: [Bug]: Headless mode awaits for requested user feedb…

### LINKED ISSUE #5015: [Bug]: Headless mode awaits for requested user feedback without showing any text for what that feedback should be
### Is there an existing issue for the same bug?

- [X] I have checked the existing issues.

### Describe the bug and reproduction steps

Command to run:

```
docker run -it \
    --pull=always \
    -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:0.13-nikolaik \
    -e SANDBOX_USER_ID=$(id -u) \
    -e WORKSPACE_MOUNT_PATH=$(pwd) \
    -e LLM_API_KEY=*** \
    -e LLM_MODEL="anthropic/claude-3-5-sonnet-20241022" \
    -v $WORKSPACE_BASE:/opt/workspace_base \
    -v /var/run/docker.sock:/var/run/docker.sock \
    --add-host host.docker.internal:host-gateway \
    --name openhands-app-$(date +%Y%m%d%H%M%S) \
    docker.all-hands.dev/all-hands-ai/openhands:0.13 \
    python -m openhands.core.main -t "the realtor_tour_scenario.test.ts is erroring, could you investigate? I think the erroring one is due to the message coming from the wrong POV" --debug true
```

output (happens twice):

```
0.13: Pulling from all-hands-ai/openhands
Digest: sha256:28307e6ef3ca477df56e0689c0fcaa8b6f073d018cf90c945317b5b76c8566cb
Status: Image is up to date for docker.all-hands.dev/all-hands-ai/openhands:0.13
Starting OpenHands...
Setting up enduser with id 501
Docker socket group id: 0
root:x:0:
Group with id 0 already exists
Running as enduser
03:16:36 - openhands:INFO: eventstream_runtime.py:220 - [runtime default-2a338e0f97031b24] Starting runtime with image: docker.all-hands.dev/all-hands-ai/runtime:0.13-nikolaik
03:16:36 - openhands:INFO: eventstream_runtime.py:224 - [runtime default-2a338e0f97031b24] Container started: openhands-runtime-default-2a338e0f97031b24
03:16:36 - openhands:INFO: eventstream_runtime.py:227 - [runtime default-2a338e0f97031b24] Waiting for client to become ready at http://host.docker.internal:38456...
03:16:54 - openhands:INFO: eventstream_runtime.py:233 - [runtime default-2a338e0f97031b24] Runtime is ready.
03:16:54 - openhands:WARNING: state.py:119 - Could not restore state from session: sessions/default-2a338e0f97031b24/agent_state.pkl
03:16:54 - openhands:INFO: agent_controller.py:193 - [Agent Controller default] Starting step loop...
03:16:54 - openhands:INFO: agent_controller.py:316 - [Agent Controller default] Setting agent(CodeActAgent) state from AgentState.LOADING to AgentState.RUNNING


==============
[Agent Controller default] LEVEL 0 LOCAL STEP 0 GLOBAL STEP 0



==============
[Agent Controller default] LEVEL 0 LOCAL STEP 1 GLOBAL STEP 1



==============
[Agent Controller default] LEVEL 0 LOCAL STEP 2 GLOBAL STEP 2

03:17:08 - openhands:INFO: agent_controller.py:316 - [Agent Controller default] Setting agent(CodeActAgent) state from AgentState.RUNNING to AgentState.AWAITING_USER_INPUT
Request user input >> the realtor_tour_scenario.test.ts is erroring, could you investigate? I think the erroring one is due to the message coming from the wrong POV, meaning the test is either using the wrong assistant, perhaps because it's not getting the right result for isRealtor                                                      
04:01:21 - openhands:INFO: agent_controller.py:316 - [Agent Controller default] Setting agent(CodeActAgent) state from AgentState.AWAITING_USER_INPUT to AgentState.RUNNING


==============
[Agent Controller default] LEVEL 0 LOCAL STEP 3 GLOBAL STEP 3

04:01:25 - openhands:INFO: agent_controller.py:316 - [Agent Controller default] Setting agent(CodeActAgent) state from AgentState.RUNNING to AgentState.AWAITING_USER_INPUT
Request user input >> ^C
Session terminated, killing shell...^C ...killed.
^C%       
```

Unsure exactly what it's awaiting request user input? it doesn't say what type of user input it wants

### OpenHands Installation

Docker command in README

### OpenHands Version

0.13

### Operating System

None

### Logs, Errors, Screenshots, and Additional Context

Executed this in the root of my typescript repo

### PR DESCRIPTION
**Pull request by enyst**
**End-user friendly description of the problem this fixes or functionality that this introduces**

- [ ] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below
Headless mode must not stop to ask the user for input.

**Explain what this PR does**

When the app is running in headless mode, it must not ask the user for input. This means two things:
- if the agent is nevertheless waiting for input, send it automatically some message, apparently from a fake "user", to force it to continue instead of waiting.
- do not raise.

Read the comments in the original issue: you will find an example of how headless mode is implemented silently in external scripts.

You have a unit test for headless mode set. Read it and fix it to conform to this behavior. Then add a test for headless mode not set.

---
**Link of any specific issues this addresses**
Fix #5015 

---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:38deae5-nikolaik   --name openhands-app-38deae5   docker.all-hands.dev/all-hands-ai/openhands:38deae5
```

### CODE REVIEW

**Comment by enyst** [comment]
The PR you are fixing is: https://github.com/All-Hands-AI/OpenHands/pull/5246

Read all its contents before starting.

**Comment by openhands-agent** [comment]
New OpenHands update

**Comment by enyst** [comment]
@openhands-agent  **End-user friendly description of the problem this fixes or functionality that this introduces:**
Headless mode must not stop to ask the user for input.

**Explain what this PR does**

When the app is running in headless mode, it must not ask the user for input. This means two things:

if the agent is nevertheless waiting for input, send it automatically some message, apparently from a fake "user", to force it to continue instead of waiting.
do not raise.
Read the comments in the original issue: you will find an example of how headless mode is implemented silently in external scripts.

You have a unit test for headless mode set. Read it and fix it to conform to this behavior. Then add a test for headless mode not set.

**Comment by openhands-agent** [comment]
New OpenHands update

**Comment by enyst** [comment]
openhands-agent, the previous comments were mostly solved. We have a few more things to do now:
- review where the AgentController is initialized, and if headless mode is set, set a fake user response there
- HOWEVER, do not overwrite a fake user response coming from a caller, if that is the case!
- verify what fake user responses exist in the codebase. Hint: ./evaluation subfolders.
- only if none is provided, and the mode is headless, have some default that is simple, and tells the agent something similar with the existing ones
- unit tests: fix the test on this issue, and add a test for when the mode isn't headless.

**Comment by openhands-agent** [comment]
New OpenHands update

**Comment by enyst** [comment]
@openhands-agent Python linting is failing on this PR in CI. You know how to run python linting in this project. Fix it and commit the fix.

**Comment by openhands-agent** [comment]
New OpenHands update

**Comment by enyst** [comment]
@openhands-agent Look carefully at the diff of this PR. Undo the exact changes you made to traffic control. Only those, and commit; the rest is fine.

**Comment by openhands-agent** [comment]
New OpenHands update

File: openhands/controller/agent_controller.py [hunk 1]
```diff
@@ -313,7 +323,25 @@ async def _handle_message_action(self, action: MessageAction) -> None:
...
+                        EventSource.USER,
+                    )
+                else:
+                    # If no fake response function is provided, we continue with an empty response
+                    self.event_stream.add_event(
+                        MessageAction(content=''),
+                        EventSource.USER,
+                    )
+            else:
+                # Display the message content to help user understand what input is expected
```
Comment (enyst) [inline openhands/controller/agent_controller.py:341]:
```suggestion
```

File: openhands/controller/agent_controller.py [hunk 2]
```diff
@@ -313,7 +323,24 @@ async def _handle_message_action(self, action: MessageAction) -> None:
...
+                        EventSource.USER,
+                    )
+                else:
+                    # If no fake response function is provided, we continue with an empty response
+                    self.event_stream.add_event(
+                        MessageAction(content=''),
+                        EventSource.USER,
+                    )
+            else:
+                print(f'\nAgent is requesting input: {action.content}')
```
Comment (enyst) [inline openhands/controller/agent_controller.py:341]:
```suggestion
```

File: openhands/controller/agent_controller.py [hunk 3]
```diff
@@ -313,7 +323,23 @@ async def _handle_message_action(self, action: MessageAction) -> None:
...
+                        EventSource.USER,
+                    )
+                else:
+                    # If no fake response function is provided, we continue with an empty response
+                    self.event_stream.add_event(
+                        MessageAction(content=''),
+                        EventSource.USER,
+                    )
+            else:
+                print('Request user input >> ', end='', flush=True)
```
Comment (enyst) [inline openhands/controller/agent_controller.py:341]:
```suggestion
```

File: tests/unit/test_agent_controller.py [hunk 4]
```diff
@@ -355,3 +355,71 @@ async def test_step_max_budget_headless(mock_agent, mock_event_stream):
...
+        confirmation_mode=False,
+        headless_mode=False,
+    )
+    controller.state.agent_state = AgentState.RUNNING
+    message_action = MessageAction(content='Test message', wait_for_response=True)
+    message_action._source = EventSource.AGENT
+    await controller.on_event(message_action)
+    # In non-headless mode, should wait for user input
+    assert controller.state.agent_state == AgentState.AWAITING_USER_INPUT
+    mock_event_stream.add_event.assert_not_called()
```
Comment (enyst) [inline tests/unit/test_agent_controller.py:424]: @openhands-agent This test is failing, and it fails here, at the line asserting that add_event is not called:

> @pytest.mark.asyncio
    async def test_message_action_user_input_non_headless(mock_agent, mock_event_stream):
        controller = AgentController(
            agent=mock_agent,
            event_stream=mock_event_stream,
            max_iterations=10,
            sid='test',
            confirmation_mode=False,
            headless_mode=False,
        )
        controller.state.agent_state = AgentState.RUNNING
        message_action = MessageAction(content='Test message', wait_for_response=True)
        message_action._source = EventSource.AGENT
        await controller.on_event(message_action)
        # In non-headless mode, should wait for user input
        assert controller.state.agent_state == AgentState.AWAITING_USER_INPUT
>       mock_event_stream.add_event.assert_not_called()

> tests/unit/test_agent_controller.py:424: 

> self = <MagicMock name='mock.add_event' id='[140554164855408](tel:140554164855408)'>
> E           AssertionError: Expected 'add_event' to not have been called. Called 1 times.
> E           Calls: [call(AgentStateChangedObservation(content='', agent_state=<AgentState.AWAITING_USER_INPUT: 'awaiting_user_input'>, observation='agent_state_changed'), <EventSource.ENVIRONMENT: 'environment'>)].

Look very carefully at the error. First, ALWAYS when an agent changes state, the app adds an AgentStateChangedObservation. Second, look at the attributes of this AgentStateChangedObservation: it says that the state of the agent is exactly AWAITING_USER_INPUT, which is what we wanted!

Fix the test.

**Comment by openhands-agent** [comment]
New OpenHands update

File: tests/unit/test_agent_controller.py (hunk 4 above)
Comment (enyst) [reply tests/unit/test_agent_controller.py]: @openhands-agent You didn't address this review comment enough, about the unit test.

The test should look at the event and its properties and verify that:
- the event is a AgentStateChangedObservation
- its agent_state attribute is AgentState.AWAITING_USER_INPUT.

**Comment by openhands-agent** [comment]
New OpenHands update

**Pull request closed without merging**